python/pyiceberg/table/__init__.py

```diff
@@ -401,9 +423,38 @@ def plan_files(self) -> Iterator[FileScanTask]:
                             metrics_evaluator,
                         )
                         for manifest in manifests
+                        if (manifest.content is None or manifest.content == ManifestContent.DATA)
+                        or (
+                            # Not interested in deletes that are older than the data
+                            manifest.content == ManifestContent.DELETES
+                            and (manifest.sequence_number or INITIAL_SEQUENCE_NUMBER) >= min_sequence_number
+                        )
                     ],
                 )
+            ):
+                if datafile.content is None or datafile.content == DataFileContent.DATA:
+                    data_datafiles.append(datafile)
+                elif datafile.content == DataFileContent.POSITION_DELETES:
+                    deletes_positional.append(datafile)
+                elif datafile.content == DataFileContent.EQUALITY_DELETES:
+                    raise ValueError(
+                        "PyIceberg does not yet support equality deletes: https://github.com/apache/iceberg/issues/6568"
+                    )
+                else:
+                    raise ValueError(f"Unknown DataFileContent: {datafile.content}")
+
+        return [
+            FileScanTask(data_file, delete_files=self._match_deletes_to_datafile(data_file, deletes_positional))
+            for data_file in data_datafiles
+        ]
+
+    def _match_deletes_to_datafile(self, data_file: DataFile, positional_delete_files: List[DataFile]) -> Set[DataFile]:
+        return set(
```

Comment:
I think this is going to over-match delete files to data files. An older delete file could match a new data file because the range of `file_path` could be large and not helpful.

I think all you need to do is to check this preliminary set by comparing the sequence number of the data file and the matching delete files.

Reply:
Ahh, I see. Yes. I agree. I also recall this from the Java implementation, but I was confused with the sequence numbers on the manifest-list level.

I've added a sorted-list for collecting the deletes by sequence number, then we can efficiently bisect them by only selecting the deletes that came after the data file.